numpy/core/_internal.py

```diff
@@ -355,15 +355,38 @@ def _view_is_safe(oldtype, newtype):
         If the new type is incompatible with the old type.
 
     """
+    from numpy.lib.type_check import find_dtype_offsets
 
     # if the types are equivalent, there is no problem.
     # for example: dtype((np.record, 'i4,i4')) == dtype((np.void, 'i4,i4'))
     if oldtype == newtype:
         return
 
-    if newtype.hasobject or oldtype.hasobject:
-        raise TypeError("Cannot change data-type for object array.")
-    return
+    # no object members is fine
+    if not newtype.hasobject and not oldtype.hasobject:
+        return
```

Comment:
This is the part I am still thinking about, becuase of the problem of "hidden objects".

It's true that currently, I don't think there is a way to create hidden objects because most operations involving objects are disallowed or make copies.

But we will have a problem when trying to solve #5994, as I attempt in my PR #6053. That is, once we allow multi-field views like:
```python
>>> a = np.zeros(5, dtype='O,O,O')
>>> b = a[['f0','f2']]
>>> b.dtype
dtype({'names':['f0','f2'], 'formats':['O','O'], 'offsets':[0,16], 'itemsize':24})
```
then b will have a hidden object at byte 8. That's will be true of any implementation of #5994, besides #6053. (Currently the second line returns a copy instead of a view, this avoiding hidden objects, but that is what we want to change).

I am trying to figure out the easiest way to get both this PR and multi-field views to interact safely. (this may involve setting limitations on the plans in #5994).

Note how this PR would allow this view, using `b` from my example above:
```
>>> b.view('O,p,O')
>>> b['f1']  = 0   # segfault?
```

Perhaps the solution is to disallow multi-field views if it would create hidden objects? That also seems a little wonky. I need to think a bit more about it.